modules/rgbd/src/warpfield.cpp

```diff
@@ -0,0 +1,335 @@
+{
+    CV_TRACE_FUNCTION();
+
+    regGraphNodes.clear();
+
+    if(n_levels == 1) // First level (Nwarp) is already stored in nodes
+    {
+        return;
+    }
+
+    float effResolution = baseRes*resGrowthRate;
+    NodeVectorType curNodes = nodes;
+    Mat curNodeMatrix = getNodesPos(curNodes);
+
+    Ptr<flann::GenericIndex<flann::L2_Simple<float> > > curNodeIndex(
+        new flann::GenericIndex<flann::L2_Simple<float> >(curNodeMatrix,
+                                                          cvflann::LinearIndexParams()));
+
+    for(int l = 0; l < (n_levels-1); l++)
+    {
+        std::vector<bool> nodeValidity(curNodeIndex->size(), true);
+        NodeVectorType coarseNodes = subsampleIndex(curNodeMatrix, *curNodeIndex,
+                                                    nodeValidity, effResolution);
+
+        Mat coarseNodeMatrix = getNodesPos(coarseNodes);
+
+        Ptr<flann::GenericIndex<flann::L2_Simple<float> > > coarseNodeIndex(
+            new flann::GenericIndex<flann::L2_Simple<float> >(coarseNodeMatrix,
+                                                              cvflann::LinearIndexParams()));
+
+        for(size_t i = 0; i < curNodes.size(); i++)
+        {
+            std::vector<int> children_indices(k);
+            std::vector<float> children_dists(k);
+
+            std::vector<float> query = {curNodeMatrix.at<float>(i, 0),
+                                        curNodeMatrix.at<float>(i, 1),
+                                        curNodeMatrix.at<float>(i, 2)};
+
+            coarseNodeIndex->knnSearch(query, children_indices, children_dists, k,
+                                       cvflann::SearchParams());
+
+            curNodes[i]->children.clear();
+            for(auto index: children_indices)
+            {
+                curNodes[i]->children.push_back(coarseNodes[index]);
+            }
+        }
+
+        regGraphNodes.push_back(coarseNodes);
+        curNodes = coarseNodes;
+        curNodeMatrix = coarseNodeMatrix;
+        curNodeIndex = coarseNodeIndex;
+        effResolution *= resGrowthRate;
+    }
+
+}
+
+Point3f WarpField::applyWarp(Point3f p, int neighbours[], int n) const
+{
+    CV_TRACE_FUNCTION();
+
+    if(n == 0)
+    {
+        return p;
+    }
+
+    float totalWeight = 0;
+    Point3f WarpedPt(0,0,0);
+
+    for(int i = 0; i < n; i++)
+    {
+        Ptr<WarpNode> neigh = nodes[neighbours[i]];
+        float w = neigh->weight(p);
+        if(w < 0.01)
+        {
+            continue;
+        }
+
+        Matx33f R = neigh->transform.rotation();
+        Vec3f T = neigh->transform.translation();
+
+        Point3f newPt = R * (p - neigh->pos) + neigh->pos;
+        newPt.x += T[0];
+        newPt.y += T[1];
+        newPt.z += T[2];
+
+        WarpedPt += newPt * w;
+        totalWeight += w;
+
+    }
+    WarpedPt /= totalWeight;
+
+    if(totalWeight == 0)
+    {
+        return cameraPoseInv * p;
+    }
+    else
+    {
+        return cameraPoseInv * WarpedPt;
```

Comment:
Warp field should not know anything about a camera which looks at the volume.
This matrix-vector multiplication should be performed at a place where this function is called. For example, in  `TSDFVolume::integrate()` function the `cameraPoseInv` can be joined together with `TSDFVolume::pose` matrix to perform just one matrix-vector multiplication (this happened before with `vol2cam`).

Reply:
Alright, I will amend this.